### PR TITLE
Add support for inhomogeneous tuple annotations

### PR DESCRIPTION
**Pull request by EmilyBourne**
Add support for inhomogeneous tuple annotations

### CODE REVIEW

File: tests/epyccel/test_epyccel_variable_annotations.py [hunk 1]
```diff
@@ -324,9 +324,41 @@ def dict_empty_init():
...
+
+def test_inhomogeneous_tuple_annotation_3(language):
+    def inhomogeneous_tuple_annotation():
+        a : tuple[int,int,int] = (1,2,3) #pylint: disable=unsubscriptable-object
+        return a[0], a[1], a[2]
+
+    epyc_inhomogeneous_tuple_annotation = epyccel(inhomogeneous_tuple_annotation, language = language)
+    assert epyc_inhomogeneous_tuple_annotation() == inhomogeneous_tuple_annotation()
+
+def test_inhomogeneous_tuple_annotation_4(language):
```
Comment (jalalium) [inline tests/epyccel/test_epyccel_variable_annotations.py:358]: The following test raises errors in both `C` and `Fortran`:

```Python
if __name__ == '__main__':
    a : tuple[tuple[int, float]] = ((1,0.2),)
    print(a[0][0], a[0][1])
```

```bash

ERROR at annotation (semantic) stage
pyccel:
 |error [semantic]: boo.py [3,4]| Cannot modify 'const' variable (a)
 ```

**Review by jalalium — CHANGES_REQUESTED** [review]

File: tests/epyccel/test_epyccel_variable_annotations.py (hunk 1 above)
Comment (EmilyBourne) [reply tests/epyccel/test_epyccel_variable_annotations.py:358]: Fixed over [1f4b1b6](https://github.com/pyccel/pyccel/pull/2066/commits/1f4b1b6a3cd039521b554044618593a3e11e519c) - [bcd4a6c](https://github.com/pyccel/pyccel/pull/2066/commits/bcd4a6c66d3eeb84712dd64b40388611debfa755)

Comment (EmilyBourne) [reply tests/epyccel/test_epyccel_variable_annotations.py:358]: Test added here : 413b93ec6

**Review by jalalium — CHANGES_REQUESTED** [review]
there are still issues with nested tuples

File: pyccel/parser/semantic.py [hunk 2]
```diff
@@ -1709,6 +1709,13 @@ def _ensure_inferred_type_matches_existing(self, class_type, d_var, var, is_auga
                 tmp_result = PyccelAdd(var, rhs)
                 result_type = tmp_result.class_type
                 raise_error = var.class_type != result_type
+            elif isinstance(var.class_type, InhomogeneousTupleType) and \
+                    isinstance(class_type, HomogeneousTupleType):
+                raise_error = any(a != class_type.element_type for a in var.class_type) or var.shape[0] != d_var['shape'][0]
```
Comment (EmilyBourne) [inline pyccel/parser/semantic.py:1714]: @jalalium these lines are the source of the errors you are seeing. Do you have a suggestion for how I should handle this? I can make a tuple with 1 element inhomogeneous by default. Usually homogenous is easier to match with everything else but not here because we also need to check zhe size. Maybe I need a recursive function

File: tests/epyccel/test_epyccel_variable_annotations.py (hunk 1 above)
Comment (EmilyBourne) [reply tests/epyccel/test_epyccel_variable_annotations.py:358]: Hmm, that's not great. I guess the infinite loop is unrelated. I'll look into it

Comment (EmilyBourne) [reply tests/epyccel/test_epyccel_variable_annotations.py:358]: Attempt number 2
Fixed?
e108ec2c5
15de72a1f
6276a382a

Comment (EmilyBourne) [reply tests/epyccel/test_epyccel_variable_annotations.py:358]: :sob: You are too good at finding these corner cases :sweat_smile:

Comment (EmilyBourne) [reply tests/epyccel/test_epyccel_variable_annotations.py:358]: Fixed : 35886366c4e9cd07b87cdcb7ee342f8ebb161084
with test : 0ad38c9ad360436aaec083ccc3c96c59fa0ca1d8

Comment (jalalium) [reply tests/epyccel/test_epyccel_variable_annotations.py:358]: Not quite 😅
```python
if __name__ == '__main__':
    a : tuple[tuple[tuple[tuple[int, float]], int]] = ((((1,0.2),),1),)
```
I think this needs changing:
```python
                    if not isinstance(new_element_type, HomogeneousTupleType):
                        raise_error = True
                        break
```

**Review by jalalium — CHANGES_REQUESTED** [review]
Some corner cases still need work

File: tests/epyccel/test_epyccel_variable_annotations.py (hunk 1 above)
Comment (EmilyBourne) [reply tests/epyccel/test_epyccel_variable_annotations.py:358]: Fixed : d7e6bbd4f
with test : a905aa787

Comment (jalalium) [reply tests/epyccel/test_epyccel_variable_annotations.py:358]: Great, this fixes the issue! but you don't try to make tuple homogeneous when possible:
```python
if __name__ == '__main__':
    b : tuple[int,int] = (1,2)
    a : tuple[tuple[tuple[tuple[int, float]], int], tuple[int, int], tuple[float,float]] = ((((1,0.2),),1),b,(0.3,0.2))
```
```C
#include "boo.h"
#include <stdlib.h>
#include <stdint.h>
int main()
{
    int64_t b_0;
    int64_t b_1;
    int64_t a_0_0_0_0;
    double a_0_0_0_1;
    int64_t a_0_1;
    int64_t a_1_0;
    int64_t a_1_1;
    double a_2_0;
    double a_2_1;
    b_0 = INT64_C(1);
    b_1 = INT64_C(2);
    a_0_0_0_0 = INT64_C(1);
    a_0_0_0_1 = 0.2;
    a_0_1 = INT64_C(1);
    a_1_0 = b_0;
    a_1_1 = b_1;
    a_2_0 = 0.3;
    a_2_1 = 0.2;
    return 0;
}
```
Did you have this in mind?

Comment (EmilyBourne) [reply tests/epyccel/test_epyccel_variable_annotations.py:358]: I have made this more explicit in the docs: f28c641bd

Comment (jalalium) [reply tests/epyccel/test_epyccel_variable_annotations.py:358]: Got it! Awesome

**Review by jalalium — APPROVED** [review]
Well done!

File: pyccel/parser/semantic.py [hunk 3]
```diff
@@ -1709,6 +1709,26 @@ def _ensure_inferred_type_matches_existing(self, class_type, d_var, var, is_auga
...
                 raise_error = var.class_type != result_type
+            elif isinstance(var.class_type, InhomogeneousTupleType) and \
+                    isinstance(class_type, HomogeneousTupleType):
+                if d_var['shape'][0] == var.shape[0]:
+                    rhs_elem = self.scope.collect_tuple_element(var[0])
+                    self._ensure_inferred_type_matches_existing(class_type.element_type,
+                            self._infer_type(rhs_elem), rhs_elem, is_augassign, new_expressions, rhs)
+                    raise_error = False
+                else:
+                    raise_error = True
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/parser/semantic.py:1720]: /bot accept
Error handling
